src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1771,22 +1864,103 @@ public sealed class One : Tests
                     builder.CheckConsistentDecision(
                         test: test,
                         other: Test,
+                        state: state,
                         whenTrueValues: whenTrueValues,
                         whenFalseValues: whenFalseValues,
                         syntax: test.Syntax,
+                        lengthTest: out Tests lengthTest,
                         trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
                         falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
                         trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
                         falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
                         foundExplicitNullTest: ref foundExplicitNullTest);
-                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
-                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    whenTrue = trueDecisionImpliesTrueOther ? makeOrSequence(lengthTest, this) : trueDecisionPermitsTrueOther ? this : makeAndSequence(Not.Create(lengthTest), this);
```

Comment:
This is where the magic happens. Still trying to wrap my head around all the cases.

Could use an explanatory comment ;-)

I'm wondering whether using `null` `lengthTest` (instead of `True`) for scenario where no addittional test is needed to compute the four flags (trueTestPermitsTrueOther/falseTestPermitsTrueOther/trueTestImpliesTrueOther/falseTestImpliesTrueOther) would make this easier to follow.

Reply:
>I'm wondering whether using null lengthTest 

I've used True here to show the actual logic that is happening. Using null might be not as clear. 
I can add a comment and rename the parameter (maybe `precondition`?)